include/beast/websocket/detail/utf8_checker.hpp

```diff
@@ -39,112 +43,264 @@ namespace detail {
-    while(size)
-    {
-        auto const byte = *p;
-        auto const type = plut[byte];
-        if(state_)
-            codepoint_ = (byte & 0x3fu) | (codepoint_ << 6);
-        else
-            codepoint_ = (0xff >> type) & byte;
-        state_ = plut[256 + state_ * 16 + type];
-        if(state_ == 1)
-        {
-            reset();
-            return false;
-        }
-        ++p;
-        --size;
-    }
-    return true;
+    auto const success = need_ == 0;
+    reset();
+    return success;
 }
 
 template<class _>
-template<class BufferSequence>
+template<class ConstBufferSequence>
 bool
-utf8_checker_t<_>::write(BufferSequence const& bs)
+utf8_checker_t<_>::write(ConstBufferSequence const& bs)
 {
+    static_assert(is_ConstBufferSequence<ConstBufferSequence>::value,
+        "ConstBufferSequence requirements not met");
     using boost::asio::buffer_cast;
     using boost::asio::buffer_size;
     for(auto const& b : bs)
-        if(! write(buffer_cast<void const*>(b),
+        if(! write(buffer_cast<std::uint8_t const*>(b),
                 buffer_size(b)))
             return false;
     return true;
 }
 
 template<class _>
 bool
-utf8_checker_t<_>::finish()
+utf8_checker_t<_>::write(std::uint8_t const* in, std::size_t size)
 {
-    auto const success = state_ == 0;
-    reset();
-    return success;
+    auto const valid =
+        [](std::uint8_t const*& in)
+        {
+            if (in[0] < 128)
+            {
+                ++in;
+                return true;
+            }
+            if ((in[0] & 0x60) == 0x40)
+            {
+                if ((in[0] & 0xfe) == 0xc0 ||
+                    (in[1] & 0xc0) != 0x80)
+                        return false;
+                in += 2;
+                return true;
+            }
+            if ((in[0] & 0xf0) == 0xe0)
+            {
+                if ((in[1] & 0xc0) != 0x80 ||
+                    (in[2] & 0xc0) != 0x80 ||
+                    (in[0] == 224 && in[1] < 160) ||
+                    (in[0] == 237 && in[1] > 159))
+                        return false;
+                in += 3;
+                return true;
+            }
+            if ((in[0] & 0xf8) == 0xf0)
+            {
+                if (in[0] > 244 ||
+                    (in[1] & 0xc0) != 0x80 ||
+                    (in[2] & 0xc0) != 0x80 ||
+                    (in[3] & 0xc0) != 0x80 ||
+                    (in[0] == 240 && in[1] < 144) ||
+                    (in[0] == 244 && in[1] > 143))
+                        return false;
+                in += 4;
+                return true;
+            }
+            return false;
+        };
+    auto const valid_have =
+        [&]()
+        {
+            if ((have_[0] & 0x60) == 0x40)
+                return have_[0] <= 223;
+            if ((have_[0] & 0xf0) == 0xe0)
+            {
+                if (p_ - have_ > 1 &&
+                    ((have_[1] & 0xc0) != 0x80 ||
+                    (have_[0] == 224 && have_[1] < 160) ||
```

Comment:
I don't know about all that, further subdivision into lambdas would make it harder to read.